README.md
# Cordova-camera-plugin-debug
A plugin to help with debugging Cordova Camera Plugin in Chrome.

## How it works


CHANGELOG.md
### Changelog
All notable changes to this project will be documented in this file.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [v1.0.3](https://github.com/michogar/cordova-camera-plugin-debug/compare/v1.0.1...v1.0.3)
> 23 January 2018
- If no camera available, degrade gracefully. Bump version. [`#11`](https://github.com/michogar/cordova-camera-plugin-debug/pull/11)
- If no camera available, degrade gracefully [`13af44b`](https://github.com/michogar/cordova-camera-plugin-debug/commit/13af44b5a96d6458ae2fe7678abfcf41b56a4e19)
- Added auto-changelog tool and CHANGELOG file [`cc50acf`](https://github.com/michogar/cordova-camera-plugin-debug/commit/cc50acf1b836da83ece921005ecf8d6d729cb918)
- Created version task with auto-changelog [`fa15908`](https://github.com/michogar/cordova-camera-plugin-debug/commit/fa1590828396d21fa03212460bd48e2e489b0865)

#### [v1.0.1](https://github.com/michogar/cordova-camera-plugin-debug/compare/v1.0.0...v1.0.1)
> 13 July 2017
- Added message in case not find camera in navigator [`f88f95e`](https://github.com/michogar/cordova-camera-plugin-debug/commit/f88f95eeb4b8184cc06be0a871c25827c3c07f7d)

### [v1.0.0](https://github.com/michogar/cordova-camera-plugin-debug/compare/v0.3.1...v1.0.0)
> 13 July 2017
- Modified the instantiation of camera amulating a polyfill [`#10`](https://github.com/michogar/cordova-camera-plugin-debug/pull/10)
- Reverted remove dist form .gitignore [`3ff6fab`](https://github.com/michogar/cordova-camera-plugin-debug/commit/3ff6fabb647120c6a3ea57779503ca8a16d4547c)

#### [v0.3.1](https://github.com/michogar/cordova-camera-plugin-debug/compare/v0.3.0...v0.3.1)
> 11 January 2017
- Created a minimal README [`#7`](https://github.com/michogar/cordova-camera-plugin-debug/pull/7)
- Removed dist folder from .gitignore to download in npm install [`cfc11f5`](https://github.com/michogar/cordova-camera-plugin-debug/commit/cfc11f568c355673982076ccdd3b329d0528813b)

#### [v0.3.0](https://github.com/michogar/cordova-camera-plugin-debug/compare/v0.2.1...v0.3.0)
> 11 January 2017
- Closes #5  [`#6`](https://github.com/michogar/cordova-camera-plugin-debug/pull/6)
- Now blob is saved as file in the filesystem and return its URL [`bd044f6`](https://github.com/michogar/cordova-camera-plugin-debug/commit/bd044f69d4c61f382f70e7bdb98f86ca56669bda)
- Released v0.3.0 [`41661fb`](https://github.com/michogar/cordova-camera-plugin-debug/commit/41661fb26955c9b6c4c92a526556f3666544dc34)

#### [v0.2.1](https://github.com/michogar/cordova-camera-plugin-debug/compare/v0.2.0...v0.2.1)
> 10 January 2017
- Repaired removeCamera method [`e051bc2`](https://github.com/michogar/cordova-camera-plugin-debug/commit/e051bc2b7eaf26a0dd8e051f1eca258fb9d56606)

#### [v0.2.0](https://github.com/michogar/cordova-camera-plugin-debug/compare/v0.1.2...v0.2.0)
> 10 January 2017
- Closes #2 [`#4`](https://github.com/michogar/cordova-camera-plugin-debug/pull/4)
- Released v0.2.0 [`a5cd404`](https://github.com/michogar/cordova-camera-plugin-debug/commit/a5cd404fb819fb90c094bf899cb94e3592019e37)
- Created modal dialog [`9a517c0`](https://github.com/michogar/cordova-camera-plugin-debug/commit/9a517c00ab0f13ea00bf4122d193a30373406558)
- Added vex library [`cf51d73`](https://github.com/michogar/cordova-camera-plugin-debug/commit/cf51d738ebcd0ae51488f340b664f54783edbccb)

#### [v0.1.2](https://github.com/michogar/cordova-camera-plugin-debug/compare/v0.1.1...v0.1.2)
> 10 January 2017
- Modified main in package.json to point to bundle.js file [`b5b42cd`](https://github.com/michogar/cordova-camera-plugin-debug/commit/b5b42cdf16f2b70ee80c9ab271344f64eb320aaf)

#### [v0.1.1](https://github.com/michogar/cordova-camera-plugin-debug/compare/v0.1.0...v0.1.1)
> 9 January 2017
- Released v0.1.1 [`3e6deca`](https://github.com/michogar/cordova-camera-plugin-debug/commit/3e6deca0b740af824edff6245c26191763fe06be)
- Added library and librayTarget to webpack to build as a library [`f4c872e`](https://github.com/michogar/cordova-camera-plugin-debug/commit/f4c872eeb1f4f3867dfccd2bdf5f0ee47b232e7c)
- Created webpack config to minify [`d70589a`](https://github.com/michogar/cordova-camera-plugin-debug/commit/d70589a1a7f628009b8c0dd425301840a1a5394b)

#### v0.1.0
> 9 January 2017
- Added camera to navigator.camera [`#1`](https://github.com/michogar/cordova-camera-plugin-debug/pull/1)
- Some minor fixes [`6fc2548`](https://github.com/michogar/cordova-camera-plugin-debug/commit/6fc25484a74a853019058427edd42f86ce813d4d)
- First commit [`ce008a7`](https://github.com/michogar/cordova-camera-plugin-debug/commit/ce008a79db0c27e8a98d4d1bc5cb6c94cd6e45d8)
- Release v0.1.0 [`d1e8c11`](https://github.com/michogar/cordova-camera-plugin-debug/commit/d1e8c11c45d017dcceb9e4e8831c7c39fa5d7c47)



tests/DummySpec.js
/**
 * Created by michogarcia on 29/12/16.
 */

import Dummy from '../src'

describe('Dummy', () => {
  it('should be dumb', () => {
    const dummy = new Dummy()
    expect(dummy.AmIDumb).toBeTruthy()
  })
})


tests/indexSpec.js
navigator.camera = undefined
const camera = require('../src') // Using require inside methods returns an empty object

describe('Camera instantiation', () => {
  it('Should camera be instantiated if navigator.camera === undefined', () => {
    expect(navigator.camera).not.toBe(undefined)
  })
})

tests/cameraSpec.js
/**
 * Created by michogarcia on 29/12/16.
 */
import Camera from '../src/Camera'

describe('Camera', () => {
  beforeEach(() => {
    new Camera()
  })

  it('should not instantiate without media', () => {
    let camera
    navigator.camera = undefined
    camera = new Camera()
    expect(navigator.camera).toBeDefined()

    camera = undefined
    try {
      camera = new Camera()
    } catch (error) {
      expect(error).not.toBeDefined()
    }
    expect(camera).toBeDefined()
  })

  it('should save blob as file after its creation', (done) => {
    const encoding = 'image/jpeg'
    const blob = new Blob([], { type: encoding })

    navigator.camera.saveFile(blob, encoding).then((uri) => {
      window.resolveLocalFileSystemURL = window.resolveLocalFileSystemURL || window.webkitResolveLocalFileSystemURL
      window.resolveLocalFileSystemURL(uri, (fileEntry) => {
        const savedURL = fileEntry.toURL()
        expect(uri).toEqual(savedURL)
        fileEntry.remove(() => {
          console.info('Dummy file removed!!')
        }, fail)
        done()
      }, fail)
    }).catch(fail)
  })

  it('should show dialog with camera', (done) => {
    navigator.camera.showCamera().then(() => {
      const video = document.getElementById('video')
      expect(video).not.toBeNull()
      expect(video.getAttribute('id')).toBe('video')
      expect(video.getAttribute('width')).toBe('400')

      const snap = document.getElementById('snap')
      expect(snap).not.toBeNull()
      expect(snap.getAttribute('id')).toBe('snap')

      const canvas = document.getElementById('canvas')
      expect(canvas).not.toBeNull()
      expect(canvas.getAttribute('id')).toBe('canvas')
      done()
    }).catch(() => {
      expect(true).toBeFalsy()
      done()
    })
  })

  it('should get encoding in mimetype way', () => {
    expect(Camera.getEncodingType(Camera.EncodingType.JPEG)).toBe('image/jpeg')
    expect(Camera.getEncodingType(Camera.EncodingType.PNG)).toBe('image/png')
    let encoding = null
    try {
      const NO_ENCODING_TYPE = 3
      encoding = Camera.getEncodingType(NO_ENCODING_TYPE)
    } catch (err) {
      expect(err.message).toBe('Not encoding allowed')
      expect(encoding).toBeNull()
    }
  })

  it('should get a picture from camera', () => {
    const canvas = document.getElementById('canvas')
    spyOn(canvas, 'toBlob')
    spyOn(canvas, 'toDataURL')
    navigator.camera.getPictureFromCamera({
      sourceType: Camera.PictureSourceType.CAMERA,
      destinationType: Camera.DestinationType.FILE_URI,
      quality: 50
    })
    expect(canvas.toBlob).toHaveBeenCalled()

    navigator.camera.getPictureFromCamera({
      sourceType: Camera.PictureSourceType.CAMERA,
      destinationType: Camera.DestinationType.DATA_URL,
      quality: 50
    })
    expect(canvas.toDataURL).toHaveBeenCalled()
  })

  it('should remove dialog with camera', (done) => {
    navigator.camera.removeCamera()

    const video = document.getElementById('video')
    expect(video).toBeNull()
    const snap = document.getElementById('snap')
    expect(snap).toBeNull()
    const canvas = document.getElementById('canvas')
    expect(canvas).toBeNull()
    done()
  })
})


karma.conf.js
// Karma configuration
// Generated on Thu Dec 29 2016 16:31:20 GMT+0100 (CET)

module.exports = function(config) {
  config.set({

    // base path that will be used to resolve all patterns (eg. files, exclude)
    basePath: '',

    webpack: { //kind of a copy of your webpack config
      devtool: 'inline-source-map', //just do inline source maps instead of the default
      module: {
        preLoaders: [
          {test: /\.js?$/, loader: 'eslint-loader', exclude: /node_modules/}
        ],
        loaders: [
          { test: /\.js$/, loader: 'babel-loader' },
          { test: /\.css$/, loader: "style-loader!css-loader" }
        ],
        postLoaders: [ { //delays coverage til after tests are run, fixing transpiled source coverage error
          test: /\.js$/,
          exclude: /(tests|node_modules|bower_components)\//,
          loader: 'istanbul-instrumenter' } ]
      }
    },

    webpackServer: {
      noInfo: true //please don't spam the console when running in karma!
    },

    coverageReporter: {
      type: 'html', //produces a html document after code is run
      dir: 'coverage/' //path to created html doc
    },

    // frameworks to use
    // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
    frameworks: ['jasmine'],


    // list of files / patterns to load in the browser
    files: [
      'webpack.config.tests.js'
    ],


    // list of files to exclude
    exclude: [
    ],


    // preprocess matching files before serving them to the browser
    // available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
    preprocessors: {
      'webpack.config.tests.js': [ 'webpack', 'sourcemap' ] //preprocess with webpack and our sourcemap loader
    },


    // test results reporter to use
    // possible values: 'dots', 'progress'
    // available reporters: https://npmjs.org/browse/keyword/karma-reporter
    reporters: ['progress'],


    // web server port
    port: 9876,


    // enable / disable colors in the output (reporters and logs)
    colors: true,


    // level of logging
    // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    logLevel: config.LOG_INFO,


    // enable / disable watching file and executing tests whenever any file changes
    autoWatch: true,


    // start these browsers
    // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
    browsers: ['Chrome'],


    // Continuous Integration mode
    // if true, Karma captures browsers, runs the tests and exits
    singleRun: false,

    // Concurrency level
    // how many browser should be started simultaneous
    concurrency: Infinity
  })
}


src/camera.js
/**
 * Created by michogarcia on 29/12/16.
 */

class Camera {

  constructor() {
    if (!navigator.camera) {
      if (!navigator.mediaDevices || !navigator.mediaDevices.getUserMedia) {
        throw new Error('No Media devices available!!')
      }
    }
  }

  showCamera() {
    return new Promise((resolve, reject) => {
      navigator.mediaDevices.getUserMedia({ video: true }).then((stream) => {
        this.video = document.createElement('video')
        this.video.setAttribute('id', 'video')
        this.video.setAttribute('width', '640')
        this.video.setAttribute('height', '480')
        this.video.setAttribute('autoplay', true)

        const snap = document.createElement('button')
        snap.setAttribute('id', 'snap')
        snap.appendChild(document.createTextNode('Snap Photo'))

        const canvas = document.createElement('canvas')
        canvas.setAttribute('id', 'canvas')
        canvas.setAttribute('width', '640')
        canvas.setAttribute('height', '480')

        document.body.appendChild(this.video)
        document.body.appendChild(snap)
        document.body.appendChild(canvas)

        this.video.src = window.URL.createObjectURL(stream)
        this.video.play()
        resolve()
      }).catch((err) => {
        this.removeCamera()
        reject(err)
      })
    })
  }

  removeCamera() {
    const snap = document.getElementById('snap')
    const canvas = document.getElementById('canvas')
    document.body.removeChild(this.video)
    document.body.removeChild(snap)
    document.body.removeChild(canvas)
  }

  getPicture(successCallback, errorCallback, options) {
    this.showCamera()
  }
}

export default Camera


src/Camera.js
/**
 * Created by michogarcia on 29/12/16.
 */
import vex from 'vex-js'
import vexDialog from 'vex-dialog'
import '../node_modules/vex-js/dist/css/vex.css'
import '../node_modules/vex-js/dist/css/vex-theme-top.css'

vex.registerPlugin(vexDialog)
vex.defaultOptions.className = 'vex-theme-top'

class Camera {

  static PictureSourceType = {
    PHOTOLIBRARY: 0,
    CAMERA: 1,
    SAVEDPHOTOALBUM: 2
  }

  static DestinationType = {
    DATA_URL: 0,
    FILE_URI: 1,
    NATIVE_URI: 2
  }

  static EncodingType = {
    JPEG: 0,
    PNG: 1
  }

  static getEncodingType(encodingType) {
    const privateEncoding = [
      'image/jpeg',
      'image/png'
    ]
    const encoding = privateEncoding[encodingType]
    if (!encoding) {
      throw new Error('Not encoding allowed')
    }
    return encoding
  }

  constructor() {
    if (!navigator.camera) {
      if (!navigator.mediaDevices || !navigator.mediaDevices.getUserMedia) {
        throw new Error('No Media devices available!!')
      } else {
        navigator.camera = this
      }
    }
  }

  showCamera() {
    return new Promise((resolve, reject) => {
      vex.dialog.open({
        message: '',
        input: [
          '<div class="vex-custom-field-wrapper">',
          '<video id="video" width="400" height="400" autoplay="true"></video>',
          '</div>',
          '<div class="vex-custom-field-wrapper">',
          '<button id="snap">Snap Photo</button>',
          '</div>',
          '<div class="vex-custom-field-wrapper">',
          '<canvas id="canvas" width="400" height="400"></canvas>',
          '</div>'
        ].join(''),
        overlayClosesOnClick: false,
        showCloseButton: false,
        escapeButtonCloses: false
      })

      navigator.mediaDevices.getUserMedia({ video: true }).then((stream) => {
        this.video = document.getElementById('video')
        this.snap = document.getElementById('snap')
        this.canvas = document.getElementById('canvas')
        this.video.src = window.URL.createObjectURL(stream)
        this.video.play()
        resolve()
      }).catch((err) => {
        this.removeCamera()
        reject(err)
      })
    })
  }

  removeCamera() {
    this.video.pause()
    this.video.parentNode.removeChild(this.video)
    this.snap.parentNode.removeChild(this.snap)
    this.canvas.parentNode.removeChild(this.canvas)
  }

  saveFile(blob, encoding) {
    const URL = window.URL || window.webkitURL
    const extension = encoding.split('/').pop()
    const name = `${URL.createObjectURL(blob).split('/').pop()}.${extension}`
    const file = new File([blob], name)

    return new Promise((resolve, reject) => {
      (window.requestFileSystem || window.webkitRequestFileSystem)(window.TEMPORARY, 10 * 1024 * 1024,
        (fs) => {
          fs.root.getFile(file.name, { create: true }, (fileEntry) => {
            fileEntry.createWriter((fileWriter) => {
              fileWriter.onwriteend = () => {
                resolve(fileEntry.toURL())
              }
              fileWriter.onerror = reject
              fileWriter.write(blob)
            }, reject)
          }, reject)
        }, reject)
    })
  }

  getPictureFromCamera({ destinationType = Camera.DestinationType.DATA_URL,
                         encodingType = Camera.EncodingType.JPEG, quality = 50 } = {}) {
    return new Promise((resolve, reject) => {
      const context = this.canvas.getContext('2d')
      context.drawImage(this.video, 0, 0, 400, 400)
      const encoding = Camera.getEncodingType(encodingType)
      if (destinationType === Camera.DestinationType.FILE_URI) {
        this.canvas.toBlob((blob) => {
          this.saveFile(blob, encoding).then(resolve).catch(reject)
        }, encoding, quality / 100)
      } else if (destinationType === Camera.DestinationType.DATA_URL) {
        resolve(this.canvas.toDataURL(encoding))
      } else {
        reject(new Error('No image!!'))
      }
    })
  }

  getPicture(successCallback, errorCallback, { sourceType = Camera.PictureSourceType.CAMERA,
    destinationType = Camera.DestinationType.DATA_URL, encodingType = Camera.EncodingType.JPEG,
    quality = 50 } = {}) {
    if (sourceType === Camera.PictureSourceType.CAMERA) {
      this.showCamera().then(() => {
        this.snap.addEventListener('click', (evt) => {
          evt.preventDefault()
          this.getPictureFromCamera({ destinationType, encodingType, quality }).then((picture) => {
            successCallback(picture)
          }).catch(errorCallback)
        })
      }).catch(errorCallback)
    }
  }
}

export default Camera
